# Introdução à Programação com Python

**Módulo:** 00 – Fundamentos e Python

**Pré-requisitos:** nenhum — este é o primeiro notebook do curso.

**Objetivos de aprendizagem:**
- Entender o que é programação e por que vamos usar Python para aprender matemática.
- Escrever e executar suas primeiras linhas de código.
- Guardar valores em variáveis e reconhecer os tipos de dados básicos (números, texto, lógico).
- Usar operadores aritméticos em Python do mesmo jeito que você usa lápis e papel.

**Tempo estimado:** 30 a 40 minutos

[![Abrir no Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/leandrofreiredealmeida/curso_matematica_python/blob/main/00_fundamentos_e_python/00_introducao_python.ipynb)

## Por que aprender a programar?

Você já usa uma calculadora. Mas uma calculadora só faz uma conta de cada vez e esquece tudo assim que você desliga: aperte `7 + 3` de novo amanhã e ela não vai lembrar de nada.

Hoje vamos construir algo diferente: um jeito de guardar números, repetir cálculos e nunca esquecer o que já foi feito. Isso, no fundo, é o que significa **programar**.

*E se você quisesse calcular o troco de 50 compras diferentes, uma atrás da outra, sem digitar a mesma conta 50 vezes? Ao final deste notebook você já vai ter as primeiras ferramentas para responder isso.*

## O que é programação e o que é Python

- Um **programa** é uma sequência de instruções que o computador executa, uma depois da outra, exatamente na ordem em que foram escritas.
- Uma **linguagem de programação** é o conjunto de regras e palavras que usamos para escrever essas instruções de um jeito que o computador entenda.
- Vamos usar **Python** porque sua sintaxe é próxima do português/inglês do dia a dia, é uma das linguagens mais usadas em ciência, matemática e análise de dados, e tem uma comunidade enorme por trás.

📌 **Definição formal:**
- Wikipédia (PT): ["Programação de computadores"](https://pt.wikipedia.org/wiki/Programa%C3%A7%C3%A3o_de_computadores)
- Wikipédia (PT): ["Python"](https://pt.wikipedia.org/wiki/Python)

**Curiosidade histórica:** Python foi criado por Guido van Rossum e batizado em homenagem ao grupo de comédia britânico *Monty Python* — não à cobra.
- Wikipédia (PT): ["Guido van Rossum"](https://pt.wikipedia.org/wiki/Guido_van_Rossum)

**Perguntas de checagem:**
1. Na sua opinião, por que um programa "não esquece" os números como uma calculadora comum?
2. Você consegue imaginar outra área (fora a matemática) onde escrever um programa ajudaria?

In [1]:
# Seu primeiro programa: uma mensagem de boas-vindas
print("Olá! Este é o seu primeiro programa em Python.")
print("A partir de agora, o computador vai fazer exatamente o que voce mandar.")

Olá! Este é o seu primeiro programa em Python.
A partir de agora, o computador vai fazer exatamente o que voce mandar.


## Variáveis

- Uma **variável** é como uma "caixinha" com nome que guarda um valor para você usar depois.
- Para colocar um valor dentro da caixinha, usamos o sinal de igual `=`.
- Regras para nomear variáveis em Python:
    - não pode começar com número (`2x` é inválido, `x2` é válido);
    - não pode ter espaços (use `_` no lugar, como em `nome_completo`);
    - maiúsculas e minúsculas são diferentes (`idade` e `Idade` são duas variáveis distintas — isso se chama *case-sensitive*).

📌 **Definição formal:**
- Wikipédia (PT): ["Variável (programação)"](https://pt.wikipedia.org/wiki/Vari%C3%A1vel_(programa%C3%A7%C3%A3o))

⚠️ **Erro comum:** confundir o `=` da programação com o `=` da matemática. Em matemática, `x = x + 1` não faz sentido (nenhum número é igual a ele mesmo mais 1). Em Python, essa linha significa "pegue o valor que está em `x`, some 1, e guarde o resultado de volta em `x`" — o sinal de igual é uma atribuição, não uma igualdade.

**Pergunta de checagem:**
1. O que acontece se você usar o mesmo nome de variável duas vezes seguidas, com valores diferentes?

In [2]:
# Criando variaveis numéricas
idade = 16
preco_lanche = 12.5
desconto = 2.0

# Mostrando os valores guardados em cada variável
print(idade)
print(preco_lanche)
print(desconto)

# Reatribuindo uma variavel: o valor antigo e substituido pelo novo
preco_lanche = preco_lanche - desconto
print("Preco com desconto:", preco_lanche)

16
12.5
2.0
Preco com desconto: 10.5


## Tipos de dados básicos

| Tipo | Exemplo | O que representa |
|---|---|---|
| `int` | `16` | números inteiros |
| `float` | `12.5` | números com casas decimais |
| `str` | `"Python"` | texto (cadeia de caracteres) |
| `bool` | `True` / `False` | verdadeiro ou falso |

Para descobrir o tipo de qualquer valor, use a função `type()`.

📌 **Definição formal:**
- Wikipédia (PT): ["Tipo de dado"](https://pt.wikipedia.org/wiki/Tipo_de_dado)

**Pergunta de checagem:**
1. Qual tipo de dado você usaria para guardar o preço de um produto? E para guardar o nome de uma pessoa?

In [3]:
# Uma variável de cada tipo básico
quantidade = 4          # int
altura = 1.68            # float
nome = "Ana"              # str
esta_aprovado = True      # bool

# Mostrando o valor e o tipo de cada variável
print(quantidade, type(quantidade))
print(altura, type(altura))
print(nome, type(nome))
print(esta_aprovado, type(esta_aprovado))

4 <class 'int'>
1.68 <class 'float'>
Ana <class 'str'>
True <class 'bool'>


## Operadores aritméticos

Os operadores abaixo funcionam exatamente como na matemática que você já conhece:

| Operador | Significado | Exemplo |
|---|---|---|
| `+` | soma | `3 + 2` → `5` |
| `-` | subtração | `3 - 2` → `1` |
| `*` | multiplicação | `3 * 2` → `6` |
| `/` | divisão | `3 / 2` → `1.5` |
| `**` | potenciação | `3 ** 2` → `9` |

E dois operadores que podem ser novidades para você:

| Operador | Significado | Exemplo |
|---|---|---|
| `//` | divisão inteira (quantas vezes cabe) | `7 // 2` → `3` |
| `%` | resto da divisão (o que sobra) | `7 // 2 = 3`, sobrando `1`, então `7 % 2` → `1` |

A **ordem de precedência** é a mesma que você já usa na matemática: parênteses primeiro, depois potência, depois multiplicação/divisão, depois soma/subtração.

📌 **Definição formal:**
- Wikipédia (PT): ["Ordem das operações"](https://pt.wikipedia.org/wiki/Ordem_das_opera%C3%A7%C3%B5es)

**Isso será usado em:** praticamente todos os notebooks seguintes, especialmente em `01_conjuntos_e_funcoes/`.

**Perguntas de checagem:**
1. Quanto é `7 // 2` e quanto é `7 % 2`? Tente calcular antes de rodar o código.
2. Por que `2 + 3 * 4` não dá `20`?

In [4]:
import ipywidgets as widgets

# Este é o seu primeiro contato com interatividade no curso:
# mexendo nos sliders, o resultado das contas muda na hora.

slider_a = widgets.IntSlider(value=8, min=-20, max=20, description="a:")
slider_b = widgets.IntSlider(value=3, min=-20, max=20, description="b:")


def mostrar_operacoes(a: int, b: int) -> None:
    '''Imprime o resultado das operacoes aritmeticas basicas entre a e b.'''
    print(f"a + b  = {a + b}")
    print(f"a - b  = {a - b}")
    print(f"a * b  = {a * b}")
    if b != 0:
        print(f"a / b  = {a / b}")
        print(f"a // b = {a // b}")
        print(f"a % b  = {a % b}")
    else:
        print("a / b, a // b e a % b nao existem quando b = 0")
    print(f"a ** b = {a ** b}")


widgets.interact(mostrar_operacoes, a=slider_a, b=slider_b)

interactive(children=(IntSlider(value=8, description='a:', max=20, min=-20), IntSlider(value=3, description='b…

<function __main__.mostrar_operacoes(a: int, b: int) -> None>

## `print()` e comentários

- A função `print()` não serve só para mostrar números: ela também mostra texto e pode misturar texto com variáveis usando **f-strings** (`f"..."`).
- **Comentários** começam com `#` são ignorados pelo Python na hora de executar o código. Servem para explicar o código para você mesmo no futuro (e para outras pessoas que forem ler).

💡 **Dica:** comente o **porquê** do código, não o **o quê** — o código já mostra o que está sendo feito; o comentário deve explicar a razão por trás da escolha.

**Pergunta de checagem:**
1. O que acontece se você colocar um comentário antes de uma linha de código?

In [5]:
nome = "Ana"
idade = 16

# Usamos f-string para misturar texto com variáveis dentro do print
print(f"Olá, {nome}! Voce tem {idade} anos.")

# Multiplicamos por 12 aqui porque queremos converter anos em meses
idade_em_meses = idade * 12
print(f"Isso equivale a {idade_em_meses} meses de vida.")

Olá, Ana! Voce tem 16 anos.
Isso equivale a 192 meses de vida.


## Resumo

| Conceito | Sintaxe | Exemplo |
|---|---|---|
| Atribuição | `=` | `x = 10` |
| Inteiro | `int` | `x = 10` |
| Decimal | `float` | `x = 10.5` |
| Texto | `str` | `x = "ola"` |
| Lógico | `bool` | `x = True` |
| Descobrir o tipo | `type()` | `type(x)` |
| Soma / Subtração | `+` `-` | `3 + 2`, `3 - 2` |
| Multiplicação / Divisão | `*` `/` | `3 * 2`, `3 / 2` |
| Divisão inteira / Resto | `//` `%` | `7 // 2`, `7 % 2` |
| Potenciação | `**` | `3 ** 2` |
| Mostrar na tela | `print()` | `print("ola")` |
| Comentário | `#` | `# explicacao` |

## Exercícios

**Básico**
1. Crie variáveis com nome, idade e cidade e imprima uma frase usando as três.
2. Calcule a área de um retângulo dados a base e a altura.

**Intermediário**
1. Dado um valor em minutos, calcule quantas horas e quantos minutos inteiros ele representa (use `//` e `%`).
2. Troque os valores de duas variáveis sem usar uma terceira variável auxiliar.

**Desafio**
1. Dado um número de segundos, calcule horas, minutos e segundos restantes, formatando a saída como `HH:MM:SS`.

Resolva cada exercício na célula de código logo abaixo (substituindo as reticências `...`) e rode a célula de verificação em seguida para conferir sua resposta.

O gabarito comentado de todos os exercícios está em [`00_introducao_python_gabarito.ipynb`](00_introducao_python_gabarito.ipynb) — só dê uma olhada depois de tentar por conta própria.

In [ ]:
# Exercício Básico 1: preencha as variáveis abaixo com seus próprios dados
nome = ...
idade = ...
cidade = ...

frase = f"Meu nome é {nome}, tenho {idade} anos e moro em {cidade}."
print(frase)

In [ ]:
# Verificação
assert nome is not Ellipsis and isinstance(nome, str) and len(nome) > 0, (
    "Tente de novo, revise o passo 1: preencha a variavel 'nome' com uma string."
)
assert idade is not Ellipsis and isinstance(idade, int), (
    "Tente de novo, revise o passo 1: preencha a variavel 'idade' com um numero inteiro."
)
assert cidade is not Ellipsis and isinstance(cidade, str) and len(cidade) > 0, (
    "Tente de novo, revise o passo 1: preencha a variavel 'cidade' com uma string."
)
print("Certo!")

In [ ]:
# Exercício Básico 2: calcule a área de um retângulo
base = 5
altura = 3

area = ...  # substitua pela fórmula da área do retângulo

print(f"Area: {area}")

In [ ]:
# Verificação
assert area == base * altura, (
    "Tente de novo, revise o passo 2: area = base * altura."
)
print("Certo!")

In [ ]:
# Exercicio Intermediário 1: converta minutos em horas e minutos
minutos_totais = 135

horas = ...    # quantas horas inteiras cabem em minutos_totais
minutos = ...  # quantos minutos sobram depois de tirar as horas inteiras

print(f"{minutos_totais} minutos = {horas}h{minutos}min")

In [ ]:
# Verificação
assert horas == minutos_totais // 60, (
    "Tente de novo, revise o passo 1: use // para calcular as horas inteiras."
)
assert minutos == minutos_totais % 60, (
    "Tente de novo, revise o passo 1: use % para calcular os minutos restantes."
)
print("Certo!")

In [ ]:
# Exercício Intermediário 2: troque os valores de a e b sem usar uma terceira variavel
a = 10
b = 25

# escreva aqui a troca dos valores de a e b


print(f"a = {a}, b = {b}")

In [ ]:
# Verificação
assert a == 25 and b == 10, (
    "Tente de novo, revise o passo 2: os valores de a e b devem ficar trocados (a=25, b=10)."
)
print("Certo!")

In [ ]:
# Exercício Desafio: converta segundos totais em HH:MM:SS
segundos_totais = 3725

horas = 0      # substitua pelo número de horas inteiras
minutos = 0    # substitua pelo número de minutos restantes
segundos = 0   # substitua pelo número de segundos restantes

tempo_formatado = f"{horas:02d}:{minutos:02d}:{segundos:02d}"
print(tempo_formatado)

In [ ]:
# Verificação
assert horas == segundos_totais // 3600, (
    "Tente de novo, revise o passo 1: quantas horas inteiras cabem no total de segundos?"
)
assert minutos == (segundos_totais % 3600) // 60, (
    "Tente de novo, revise o passo 2: quantos minutos sobram depois de tirar as horas?"
)
assert segundos == segundos_totais % 60, (
    "Tente de novo, revise o passo 3: quantos segundos sobram depois de tirar minutos e horas?"
)
assert tempo_formatado == "01:02:05", (
    "Tente de novo, revise a formatacao HH:MM:SS."
)
print("Certo!")

## Conexões

**Isso será usado em:**
- [01_conjuntos_e_funcoes/](../01_conjuntos_e_funcoes/) — toda a trilha depende deste notebook.
- Praticamente todos os notebooks seguintes usam variáveis, tipos de dados e operadores aritméticos.